File: packages/dotnet/OpenSkp/ObjExport.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace OpenSkp
{
    /// <summary>
    /// Wavefront OBJ and MTL material library text exporter for baked <see cref="Scene"/> objects.
    /// </summary>
    public static class ObjExport
    {
        private static string SanitizeMaterialName(string name)
        {
            if (string.IsNullOrWhiteSpace(name)) return "default_material";
            string clean = Regex.Replace(name.Trim(), @"[^\w\.-]", "_");
            return string.IsNullOrEmpty(clean) ? "default_material" : clean;
        }

        /// <summary>
        /// Convert a baked <see cref="Scene"/>'s materials into Wavefront MTL text format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <returns>The formatted Wavefront MTL text string.</returns>
        public static string ToMtl(Scene scene)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            var sb = new StringBuilder();
            sb.AppendLine("# OpenSKP MTL Material Library Export");
            sb.AppendLine($"# Materials: {scene.GltfMaterials.Count}");
            sb.AppendLine();

            for (int idx = 0; idx < scene.GltfMaterials.Count; idx++)
            {
                var mat = scene.GltfMaterials[idx] as Dictionary<string, object>;
                if (mat == null) continue;

                string rawName = mat.TryGetValue("name", out object? nameObj) && nameObj != null ? nameObj.ToString()! : $"Material_{idx}";
                string matName = SanitizeMaterialName(rawName);

                float r = 0.8f, g = 0.8f, b = 0.8f, a = 1.0f;
                if (mat.TryGetValue("pbrMetallicRoughness", out object? pbrObj) && pbrObj is Dictionary<string, object> pbr)
                {
                    if (pbr.TryGetValue("baseColorFactor", out object? colObj))
                    {
                        if (colObj is List<float> floatList && floatList.Count >= 3)
                        {
                            r = floatList[0]; g = floatList[1]; b = floatList[2];
                            if (floatList.Count >= 4) a = floatList[3];
                        }
                        else if (colObj is List<object> objList && objList.Count >= 3)
                        {
                            r = Convert.ToSingle(objList[0], CultureInfo.InvariantCulture);
                            g = Convert.ToSingle(objList[1], CultureInfo.InvariantCulture);
                            b = Convert.ToSingle(objList[2], CultureInfo.InvariantCulture);
                            if (objList.Count >= 4) a = Convert.ToSingle(objList[3], CultureInfo.InvariantCulture);
                        }
                    }
                }

                sb.AppendLine($"newmtl {matName}");
                sb.AppendLine("Ka 1.000000 1.000000 1.000000");
                sb.AppendLine(string.Format(CultureInfo.InvariantCulture, "Kd {0:F6} {1:F6} {2:F6}", r, g, b));
                sb.AppendLine("Ks 0.200000 0.200000 0.200000");
                sb.AppendLine("Ns 32.000000");
                sb.AppendLine(string.Format(CultureInfo.InvariantCulture, "d {0:F6}", a));
                sb.AppendLine("illum 2");

                if (mat.TryGetValue("texture_path", out object? texPathObj) && texPathObj != null)
                {
                    string texName = Path.GetFileName(texPathObj.ToString()!);
                    sb.AppendLine($"map_Kd {texName}");
                }

                sb.AppendLine();
            }

            return sb.ToString();
        }

        /// <summary>
        /// Convert a baked <see cref="Scene"/> into Wavefront OBJ text format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="mtlFilename">Optional companion .mtl filename to reference.</param>
        /// <returns>The formatted Wavefront OBJ text string.</returns>
        public static string ToObj(Scene scene, string? mtlFilename = null)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            var sb = new StringBuilder();
            sb.AppendLine("# OpenSKP OBJ Export");
            sb.AppendLine($"# Primitives: {scene.GlbPrimitives.Count}");

            if (!string.IsNullOrEmpty(mtlFilename))
            {
                sb.AppendLine($"mtllib {mtlFilename}");
            }

            sb.AppendLine();

            int vertOffset = 1;
            int uvOffset = 1;
            int normOffset = 1;

            foreach (var prim in scene.GlbPrimitives)
            {
                sb.AppendLine($"o {prim.GeomName}");

                int vertCount = prim.Positions.Length / 3;
                for (int i = 0; i < vertCount; i++)
                {
                    string x = prim.Positions[i * 3].ToString("F6", CultureInfo.InvariantCulture);
                    string y = prim.Positions[i * 3 + 1].ToString("F6", CultureInfo.InvariantCulture);
                    string z = prim.Positions[i * 3 + 2].ToString("F6", CultureInfo.InvariantCulture);
                    sb.AppendLine($"v {x} {y} {z}");
                }

                int uvCount = prim.Uvs != null ? prim.Uvs.Length / 2 : 0;
                for (int i = 0; i < uvCount; i++)
                {
                    string u = prim.Uvs![i * 2].ToString("F6", CultureInfo.InvariantCulture);
                    string v = prim.Uvs![i * 2 + 1].ToString("F6", CultureInfo.InvariantCulture);
                    sb.AppendLine($"vt {u} {v}");
                }

                int normCount = prim.Normals != null ? prim.Normals.Length / 3 : 0;
                for (int i = 0; i < normCount; i++)
                {
                    string nx = prim.Normals![i * 3].ToString("F6", CultureInfo.InvariantCulture);
                    string ny = prim.Normals![i * 3 + 1].ToString("F6", CultureInfo.InvariantCulture);
                    string nz = prim.Normals![i * 3 + 2].ToString("F6", CultureInfo.InvariantCulture);
                    sb.AppendLine($"vn {nx} {ny} {nz}");
                }

                int matIdx = prim.MaterialIndex;
                if (matIdx >= 0 && matIdx < scene.GltfMaterials.Count)
                {
                    var mat = scene.GltfMaterials[matIdx] as Dictionary<string, object>;
                    if (mat != null)
                    {
                        string matRaw = mat.TryGetValue("name", out object? nameObj) && nameObj != null ? nameObj.ToString()! : $"Material_{matIdx}";
                        sb.AppendLine($"usemtl {SanitizeMaterialName(matRaw)}");
                    }
                }

                int triCount = prim.Indices.Length / 3;
                bool hasUvs = uvCount == vertCount;
                bool hasNormals = normCount == vertCount;

                for (int i = 0; i < triCount; i++)
                {
                    uint i0 = prim.Indices[i * 3];
                    uint i1 = prim.Indices[i * 3 + 1];
                    uint i2 = prim.Indices[i * 3 + 2];

                    long v0 = i0 + vertOffset;
                    long v1 = i1 + vertOffset;
                    long v2 = i2 + vertOffset;

                    if (hasUvs && hasNormals)
                    {
                        long vt0 = i0 + uvOffset;
                        long vt1 = i1 + uvOffset;
                        long vt2 = i2 + uvOffset;
                        long vn0 = i0 + normOffset;
                        long vn1 = i1 + normOffset;
                        long vn2 = i2 + normOffset;
                        sb.AppendLine($"f {v0}/{vt0}/{vn0} {v1}/{vt1}/{vn1} {v2}/{vt2}/{vn2}");
                    }
                    else if (hasUvs)
                    {
                        long vt0 = i0 + uvOffset;
                        long vt1 = i1 + uvOffset;
                        long vt2 = i2 + uvOffset;
                        sb.AppendLine($"f {v0}/{vt0} {v1}/{vt1} {v2}/{vt2}");
                    }
                    else if (hasNormals)
                    {
                        long vn0 = i0 + normOffset;
                        long vn1 = i1 + normOffset;
                        long vn2 = i2 + normOffset;
                        sb.AppendLine($"f {v0}//{vn0} {v1}//{vn1} {v2}//{vn2}");
                    }
                    else
                    {
                        sb.AppendLine($"f {v0} {v1} {v2}");
                    }
                }

                vertOffset += vertCount;
                if (hasUvs) uvOffset += uvCount;
                if (hasNormals) normOffset += normCount;

                sb.AppendLine();
            }

            return sb.ToString();
        }

        /// <summary>
        /// Export a baked <see cref="Scene"/> directly to a Wavefront OBJ file at <paramref name="outputPath"/> and optional companion MTL file.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="outputPath">Destination file path (.obj).</param>
        /// <param name="exportMtl">Whether to export companion .mtl file alongside .obj.</param>
        public static void ExportObj(Scene scene, string outputPath, bool exportMtl = true)
        {
            if (string.IsNullOrEmpty(outputPath)) throw new ArgumentException("Output path cannot be null or empty", nameof(outputPath));

            string? mtlName = exportMtl ? $"{Path.GetFileNameWithoutExtension(outputPath)}.mtl" : null;
            string text = ToObj(scene, mtlName);
            string? dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }
            File.WriteAllText(outputPath, text, new UTF8Encoding(false));

            if (exportMtl && !string.IsNullOrEmpty(mtlName))
            {
                string mtlPath = Path.Combine(dir ?? ".", mtlName);
                File.WriteAllText(mtlPath, ToMtl(scene), new UTF8Encoding(false));
            }
        }
    }
}

File: packages/dart/lib/src/obj_export.dart
import 'dart:io';

import 'scene.dart';

String _sanitizeMaterialName(String name) {
  final clean = name.trim().replaceAll(RegExp(r'[^\w\.-]'), '_');
  return clean.isEmpty ? 'default_material' : clean;
}

/// Convert a baked [Scene]'s materials into Wavefront MTL text representation.
String toMtl(Scene scene) {
  final buffer = StringBuffer();
  buffer.writeln('# OpenSKP MTL Material Library Export');
  buffer.writeln('# Materials: ${scene.gltfMaterials.length}');
  buffer.writeln();

  for (var idx = 0; idx < scene.gltfMaterials.length; idx++) {
    final mat = scene.gltfMaterials[idx];
    final rawName = mat['name'] as String? ?? 'Material_$idx';
    final matName = _sanitizeMaterialName(rawName);

    final pbr = mat['pbrMetallicRoughness'] as Map<String, dynamic>? ?? {};
    final baseColor = (pbr['baseColorFactor'] as List?)?.cast<num>() ?? [0.8, 0.8, 0.8, 1.0];
    final r = (baseColor.isNotEmpty ? baseColor[0] : 0.8).toDouble().toStringAsFixed(6);
    final g = (baseColor.length > 1 ? baseColor[1] : 0.8).toDouble().toStringAsFixed(6);
    final b = (baseColor.length > 2 ? baseColor[2] : 0.8).toDouble().toStringAsFixed(6);
    final a = (baseColor.length > 3 ? baseColor[3] : 1.0).toDouble().toStringAsFixed(6);

    buffer.writeln('newmtl $matName');
    buffer.writeln('Ka 1.000000 1.000000 1.000000');
    buffer.writeln('Kd $r $g $b');
    buffer.writeln('Ks 0.200000 0.200000 0.200000');
    buffer.writeln('Ns 32.000000');
    buffer.writeln('d $a');
    buffer.writeln('illum 2');

    final texturePath = mat['texture_path'] as String?;
    if (texturePath != null) {
      final texName = texturePath.split(RegExp(r'[/\\]')).last;
      buffer.writeln('map_Kd $texName');
    }

    buffer.writeln();
  }

  return buffer.toString();
}

/// Convert a baked [Scene] into Wavefront OBJ text representation.
String toObj(Scene scene, {String? mtlFilename}) {
  final buffer = StringBuffer();
  buffer.writeln('# OpenSKP OBJ Export');
  buffer.writeln('# Primitives: ${scene.glbPrimitives.length}');
  if (mtlFilename != null && mtlFilename.isNotEmpty) {
    buffer.writeln('mtllib $mtlFilename');
  }
  buffer.writeln();

  var vertOffset = 1;
  var uvOffset = 1;
  var normOffset = 1;

  for (final prim in scene.glbPrimitives) {
    buffer.writeln('o ${prim.geomName}');

    final vertCount = prim.positions.length ~/ 3;
    for (var i = 0; i < vertCount; i++) {
      final x = prim.positions[i * 3].toStringAsFixed(6);
      final y = prim.positions[i * 3 + 1].toStringAsFixed(6);
      final z = prim.positions[i * 3 + 2].toStringAsFixed(6);
      buffer.writeln('v $x $y $z');
    }

    final uvs = prim.uvs;
    final uvCount = uvs.isEmpty ? 0 : uvs.length ~/ 2;
    for (var i = 0; i < uvCount; i++) {
      final u = uvs[i * 2].toStringAsFixed(6);
      final v = uvs[i * 2 + 1].toStringAsFixed(6);
      buffer.writeln('vt $u $v');
    }

    final normals = prim.normals;
    final normCount = normals.isEmpty ? 0 : normals.length ~/ 3;
    for (var i = 0; i < normCount; i++) {
      final nx = normals[i * 3].toStringAsFixed(6);
      final ny = normals[i * 3 + 1].toStringAsFixed(6);
      final nz = normals[i * 3 + 2].toStringAsFixed(6);
      buffer.writeln('vn $nx $ny $nz');
    }

    final matIdx = prim.materialIndex;
    if (matIdx >= 0 && matIdx < scene.gltfMaterials.length) {
      final matRaw = scene.gltfMaterials[matIdx]['name'] as String? ?? 'Material_$matIdx';
      buffer.writeln('usemtl ${_sanitizeMaterialName(matRaw)}');
    }

    final triCount = prim.indices.length ~/ 3;
    final hasUvs = uvCount == vertCount;
    final hasNormals = normCount == vertCount;

    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3];
      final i1 = prim.indices[i * 3 + 1];
      final i2 = prim.indices[i * 3 + 2];

      final v0 = i0 + vertOffset;
      final v1 = i1 + vertOffset;
      final v2 = i2 + vertOffset;

      if (hasUvs && hasNormals) {
        final vt0 = i0 + uvOffset;
        final vt1 = i1 + uvOffset;
        final vt2 = i2 + uvOffset;
        final vn0 = i0 + normOffset;
        final vn1 = i1 + normOffset;
        final vn2 = i2 + normOffset;
        buffer.writeln('f $v0/$vt0/$vn0 $v1/$vt1/$vn1 $v2/$vt2/$vn2');
      } else if (hasUvs) {
        final vt0 = i0 + uvOffset;
        final vt1 = i1 + uvOffset;
        final vt2 = i2 + uvOffset;
        buffer.writeln('f $v0/$vt0 $v1/$vt1 $v2/$vt2');
      } else if (hasNormals) {
        final vn0 = i0 + normOffset;
        final vn1 = i1 + normOffset;
        final vn2 = i2 + normOffset;
        buffer.writeln('f $v0//$vn0 $v1//$vn1 $v2//$vn2');
      } else {
        buffer.writeln('f $v0 $v1 $v2');
      }
    }

    vertOffset += vertCount;
    if (hasUvs) uvOffset += uvCount;
    if (hasNormals) normOffset += normCount;

    buffer.writeln();
  }

  return buffer.toString();
}

/// Export a baked [Scene] to a Wavefront OBJ file at [path] and optional companion MTL file.
void exportObj(Scene scene, String path, {bool exportMtl = true}) {
  final file = File(path);
  final dir = file.parent;
  if (!dir.existsSync()) {
    dir.createSync(recursive: true);
  }

  final filename = file.uri.pathSegments.last;
  final dotIdx = filename.lastIndexOf('.');
  final stem = dotIdx != -1 ? filename.substring(0, dotIdx) : filename;
  final mtlName = exportMtl ? '$stem.mtl' : null;

  final text = toObj(scene, mtlFilename: mtlName);
  file.writeAsStringSync(text);

  if (exportMtl && mtlName != null) {
    final mtlFile = File('${dir.path}/$mtlName');
    mtlFile.writeAsStringSync(toMtl(scene));
  }
}
